index.d.ts

```diff
@@ -0,0 +1,47 @@
+import { MutableRefObject, ReactElement, Ref } from "react";
+
+import * as CSS from "csstype";
+
+export as namespace Showtime;
+
+export type PredefinedTransition = "slideFade" | "slide" | "fade" | "rise" | "scale"
+
+export interface TransitionProperties {
+    value: string;
+    duration?: CSS.Property.TransitionDuration;
+    delay?: CSS.Property.TransitionDelay;
+    easing?: CSS.Property.TransitionTimingFunction;
+}
+
+export type Transition =
+    PredefinedTransition |
+    { [key: string]: string | number | TransitionProperties };
+
+export interface ShowtimeSettings {
+    startHidden?: boolean;
+    startWithTransition?: boolean;
+    transition?: Transition;
+    showTransition?: Transition;
+    hideTransition?: Transition;
+    duration?: CSS.Property.TransitionDuration;
+    delay?: CSS.Property.TransitionDelay;
+    easing?: CSS.Property.TransitionTimingFunction;
+    showDuration?: CSS.Property.TransitionDuration;
+    showDelay?: CSS.Property.TransitionDelay;
+    showEasing?: CSS.Property.TransitionTimingFunction;
+    hideDuration?: CSS.Property.TransitionDuration;
+    hideDelay?: CSS.Property.TransitionDelay;
+    hideEasing?: CSS.Property.TransitionTimingFunction;
+}
+
+export interface ShowtimeProps extends ShowtimeSettings {
+    children: (ref: any) => JSX.Element;
```

Comment:
Yuck. For now, this uses `any` for the ref argument. I wasn't sure how to improve this with a more constrained type, because the T of React.Ref<T> should be whatever element type is returned from the method defined for children.

Reply:
So my project is a bit messed up at the moment, but I think something like this might work to at least enforce consistency of the ref type between `Showtime` and `ShowtimeProps`:
```ts
export interface ShowtimeProps<T> extends ShowtimeSettings {
    children: (ref: Ref<T>) => JSX.Element;
    show: boolean;
    onHidden?: () => void;
    onShowing?: () => void;
}

export function useShowtime<T>(
    settings: PredefinedTransition | ShowtimeSettings
): [MutableRefObject<T>, boolean, () => void, () => void];

export function Showtime<T>(
    props: ShowtimeProps<T>
): ReactElement<ShowtimeProps<T>>;
```

Reply:
🎉 that seems to work for now.